File: solutions/sis1.c

/* -------------------------------------------------------------------------
 * This program simulates a simple (s,S) inventory system using demand read  
 * from a text file.  Backlogging is permitted and there is no delivery lag.
 * The output statistics are the average demand and order per time interval
 * (they should be equal), the relative frequency of setup and the time
 * averaged held (+) and short (-) inventory levels.
 *
 * NOTE: use 0 <= MINIMUM < MAXIMUM, i.e., 0 <= s < S.
 *
 * Name              : sis1.c  (Simple Inventory System, version 1)
 * Authors           : Steve Park & Dave Geyer 
 * Language          : ANSI C 
 * Latest Revision   : 8-20-97 
 * Compile with      : gcc sis1.c
 * ------------------------------------------------------------------------- 
*/

#include <stdio.h>                        

#define FILENAME  "sis1.dat"            /* input data file                */
#define MINIMUM   20                    /* 's' inventory policy parameter */
#define MAXIMUM   80                    /* 'S' inventory policy parameter */
#define sqr(x)    ((x) * (x))

/* ======================== */
   long GetDemand(FILE *fp)            
/* ======================== */
{
  long d;

  fscanf(fp, "%ld\n", &d);
  return (d);
}

/* ============== */
   int main(void)
/* ============== */
{
  FILE *fp;                                /* input data file         */
  long index     = 0;                      /* time interval index     */
  long inventory = MAXIMUM;                /* current inventory level */
  long demand;                             /* amount of demand        */
  long order;                              /* amount of order         */
  struct {                                 /* sum of ...              */
    double setup;                          /*   setup instances       */
    double holding;                        /*   inventory held (+)    */
    double shortage;                       /*   inventory short (-)   */
    double order;                          /*   orders                */
    double demand;                         /*   demands               */
  } sum = { 0.0, 0.0, 0.0, 0.0, 0.0 };

  fp = fopen(FILENAME, "r");
  if (fp == NULL) {
    fprintf(stderr, "Cannot open input file %s\n", FILENAME);
    return (1);
  }

  while (!feof(fp)) {
    index++;
    if (inventory < MINIMUM) {             /* place an order          */
      order         = MAXIMUM - inventory;
      sum.setup++;
      sum.order    += order;
    }
    else                                   /* no order                 */
      order         = 0;                   
    inventory      += order;               /* there is no delivery lag */
    demand          = GetDemand(fp);
    sum.demand     += demand;
    if (inventory > demand) 
      sum.holding  += (inventory - 0.5 * demand);
    else {
      sum.holding  += sqr(inventory) / (2.0 * demand);
      sum.shortage += sqr(demand - inventory) / (2.0 * demand);
    }
    inventory      -= demand;
  }

  if (inventory < MAXIMUM) {               /* force the final inventory to */
    order           = MAXIMUM - inventory; /* match the initial inventory  */
    sum.setup++;
    sum.order      += order;
    inventory      += order;
  }

  printf("\nfor %ld time intervals ", index);
  printf("with an average demand of %6.2f\n", sum.demand / index);
  printf("and policy parameters (s, S) = (%d, %d)\n\n", MINIMUM, MAXIMUM);
  
  printf("   average order (o).......... = %6.2f: %f\n", sum.order / index,sum.order / index);
  printf("   setup frequency (u)........ = %6.2f: %f\n", sum.setup / index,sum.setup / index);
  printf("   average holding level (l+). = %6.2f: %f\n", sum.holding / index,sum.holding / index);
  printf("   average shortage level (l-) = %6.2f: %f\n\n", sum.shortage / index,sum.shortage / index);

  double itemCost = sum.order / index * 8000;
  double setupCost = sum.setup / index * 1000;
  double holdCost = sum.holding / index * 25;
  double shortCost = sum.shortage / index * 700.0;

  printf("   item cost ................. = %6.2f\n", itemCost);
  printf("   setup cost ................ = %6.2f\n", setupCost);
  printf("   holding cost .............. = %6.2f\n", holdCost);
  printf("   shortage cost ............. = %6.2f\n\n", shortCost);

  double dependentCost = setupCost + holdCost + shortCost;
  double totalCost = itemCost + setupCost + holdCost + shortCost;
  printf("   dependent cost............. = %6.2f\n", dependentCost);
  printf("   total cost................. = %6.2f\n", totalCost);

  fclose(fp);
  return (0);
}

File: solutions/HW1/sis1.py

# -------------------------------------------------------------------------
#  * This program simulates a simple (s,S) inventory system using demand read  
#  * from a text file.  Backlogging is permitted and there is no delivery lag.
#  * The output statistics are the average demand and order per time interval
#  * (they should be equal), the relative frequency of setup and the time
#  * averaged held (+) and short (-) inventory levels.
#  *
#  * NOTE: use 0 <= MINIMUM < MAXIMUM, i.e., 0 <= s < S.
#  *
#  * Name              : sis1.c  (Simple Inventory System, version 1)
#  * Authors           : Steve Park & Dave Geyer 
#  * Language          : ANSI C 
#  * Latest Revision   : 8-20-97 
#  * Compile with      : gcc sis1.c
#    Translated by   : Philip Steele 
#    Language        : Python 3.3
#    Latest Revision : 3/26/14
#  * ------------------------------------------------------------------------- 
#  */                     

FILENAME = "sis1.dat"            # input data file                */
MINIMUM = 20                    # 's' inventory policy parameter */
MAXIMUM = 80                    # 'S' inventory policy parameter */
import sys
import xlwt
from xlwt import Workbook
wb = Workbook()
sheet1 = wb.add_sheet('Sheet 1')

# print(int(sys.argv[1]))

class sumOf:
  setup = 0.0     #setup instances
  holding = 0.0   #inventory held (+)
  shortage = 0.0  #inventory held (-)
  order = 0.0     #orders
  demand = 0.0    #demands

for i in range(int(sys.argv[1])):
  MINIMUM = i
################################# Main Program ##############################


  index     = 0                      # time interval index     */
  inventory = MAXIMUM                # current inventory level */
  demand = -1                            # amount of demand        */
  order  = -1                              # amount of order         */
  sum = sumOf()

  #read in demands from file 
  try:
    fp = open(FILENAME, "r")
  except IOError:
    print("File not found")
    exit()

  data = []
  for line in fp:
    data.append(int(line))

  fp.close()



  for demand in data:
    index += 1

    if (inventory < MINIMUM):              # place an order          */
      order         = MAXIMUM - inventory
      sum.setup    += 1.0
      sum.order    += order
    
    else:                                   # no order                 */
      order         = 0  

    inventory      += order               # there is no delivery lag */
    #demand          = GetDemand(fp)

    sum.demand     += demand
    if (inventory > demand): 
      sum.holding  += (inventory - 0.5 * demand)
    else:
      sum.holding  += (inventory*inventory) / (2.0 * demand)
      sum.shortage += ((demand - inventory)*(demand-inventory)) / (2.0 * demand) #debug
    
    inventory      -= demand
  #EndFor


  if (inventory < MAXIMUM):               # force the final inventory to */
    order           = MAXIMUM - inventory # match the initial inventory  */
    sum.setup += 1
    sum.order      += order
    inventory      += order


  # print("\nfor {0:1d} time intervals with an average demand of {1:6.2f}".format(index,(sum.demand / index)))
  # print("and policy parameters (s, S) = ({0}, {1})\n".format(MINIMUM, MAXIMUM))
  # print("   average order ............ = {0:6.2f}: ".format(sum.order / index) + str(sum.order / index))
  # print("   setup frequency .......... = {0:6.2f}: ".format(sum.setup / index)+ str(sum.setup / index))
  # print("   average holding level .... = {0:6.2f}: ".format(sum.holding / index)+ str(sum.holding / index))
  # print("   average shortage level ... = {0:6.2f}: ".format(sum.shortage / index)+ str(sum.shortage / index))
  # print("")
  sheet1.write(i,0,i)

  itemCost = sum.order / index * 8000
  setupCost = sum.setup / index * 1000
  holdCost = sum.holding / index * 25
  shortCost = sum.shortage / index * 700.0

  # print("   item cost ................. = {0:6.2f}: ".format(itemCost) + str(itemCost))
  # print("   setup cost ................ = {0:6.2f}: ".format(setupCost)+ str(setupCost))
  # print("   holding cost .............. = {0:6.2f}: ".format(holdCost) + str(holdCost))
  # print("   shortage cost ............. = {0:6.2f}: ".format(shortCost) + str(shortCost))
  # print("")
  sheet1.write(i,2,itemCost)
  sheet1.write(i,3,setupCost)
  sheet1.write(i,4,holdCost)
  sheet1.write(i,5,shortCost)
  
  dependentCost = setupCost + holdCost + shortCost
  totalCost = itemCost + setupCost + holdCost + shortCost
  # print("   dependent cost............. = {0:6.2f}".format(dependentCost))
  # print("   total cost................. = {0:6.2f}".format(totalCost))
  # print("")
  sheet1.write(i,7,dependentCost)
  sheet1.write(i,8,totalCost)

wb.save('Assignment1.xls') 
#C output:
# for 100 time intervals with an average demand of  29.29
# and policy parameters (s, S) = (20, 80)

#    average order ............ =  29.29
#    setup frequency .......... =   0.39
#    average holding level .... =  42.40
#    average shortage level ... =   0.25
